quiche/src/lib.rs

```diff
@@ -18940,93 +18977,145 @@ mod tests {
+    }
+
+    #[rstest]
+    /// Test that PMTUD retries with smaller probe size after loss
+    fn pmtud_probe_retry_after_loss(
+        #[values("cubic", "bbr2", "bbr2_gcongestion")] cc_algorithm_name: &str,
+    ) {
+        let mut config = Config::new(crate::PROTOCOL_VERSION).unwrap();
+        config.set_cc_algorithm_name(cc_algorithm_name).unwrap();
         config
-            .set_application_protos(&[b"proto1", b"proto2"])
+            .load_cert_chain_from_pem_file("examples/cert.crt")
             .unwrap();
+        config
+            .load_priv_key_from_pem_file("examples/cert.key")
+            .unwrap();
+        config.set_application_protos(&[b"proto1"]).unwrap();
         config.verify_peer(false);
-        config.set_initial_max_data(100000);
-        config.set_initial_max_stream_data_bidi_local(100000);
-        config.set_initial_max_stream_data_bidi_remote(100000);
-        config.set_initial_max_streams_bidi(2);
-        config.set_active_connection_id_limit(4);
-        config.set_max_send_udp_payload_size(1350);
-        config.set_max_recv_udp_payload_size(1250);
+        config.set_max_send_udp_payload_size(1400);
         config.discover_pmtu(true);
 
-        // Perform initial handshake.
         let mut pipe = testing::Pipe::with_config(&mut config).unwrap();
         assert_eq!(pipe.handshake(), Ok(()));
 
-        let server_addr = testing::Pipe::server_addr();
-        let client_addr = testing::Pipe::client_addr();
-        let pid_1 = pipe
-            .server
-            .paths
-            .path_id_from_addrs(&(server_addr, client_addr))
-            .expect("no such path");
-
-        // Check that PMTU params are configured correctly
-        let pmtu_param = &mut pipe.server.paths.get_mut(pid_1).unwrap().pmtud;
-        assert!(pmtu_param.get_should_probe());
-        assert_eq!(pmtu_param.get_probe_size(), 1350);
-        std::thread::sleep(
-            pipe.server.paths.get_mut(pid_1).unwrap().recovery.rtt() +
-                time::Duration::from_millis(1),
-        );
+        // Get initial probe size
+        let active_path = pipe.client.paths.get_active_mut().unwrap();
+        let initial_probe_size = active_path.pmtud.get_probe_size();
+        assert_eq!(initial_probe_size, 1400);
+
+        // Send first probe
+        let mut out = [0; 4096];
+        while let Ok(_) = pipe.client.send(&mut out) {}
+
+        // Verify probe flag was reset after sending
+        let active_path = pipe.client.paths.get_active_mut().unwrap();
+        assert!(!active_path.pmtud.get_should_probe());
+
+        // Simulate probe loss
+        active_path.pmtud.pmtu_probe_lost();
+
+        // Verify probe flag is re-enabled and size is reduced
+        assert!(active_path.pmtud.get_should_probe());
+        let second_probe_size = active_path.pmtud.get_probe_size();
+        assert!(second_probe_size < initial_probe_size);
+        assert!(second_probe_size > 1200);
```

Comment:
We could do an exact comparison here.  We expect the second probe to be 1300 bytes.

assert_eq!(second_probe_size, 1300);

Reply:
Done